Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -84,28 +78,25 @@ public function findByIdentifier($identifier)
     }
 
     /**
-     * @param  AggregateRootInterface $aggregate
+     * @param AggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    public function save(AggregateRootInterface $aggregate)
+    final public function save(AggregateRootInterface $aggregate, int $expectedVersion = null)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
+        if (!$aggregate instanceof EventSourcedAggregateRootInterface) {
+            throw new \InvalidArgumentException(sprintf('This repository can only save instances implementing EventSourcedAggregateRootInterface, given: "%s"', get_class($aggregate)), 1477226552);
         }
-
+        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getVersion() + 1;
```

Comment:
>  it needs to stay consistent at all times

Maybe the naming is confusing, but the version should reflect the version of the aggregate in it's stream. It can have uncommitted events on top of this, the version is still the same.
So in fact, we should *not* allow the version to be changed from within the aggregate, if possible